sentry-ruby/lib/sentry/propagation_context.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+require "securerandom"
+require "sentry/baggage"
+
+module Sentry
+  class PropagationContext
+
+    # An uuid that can be used to identify a trace.
+    # @return [String]
+    attr_reader :trace_id
+    # An uuid that can be used to identify the span.
+    # @return [String]
+    attr_reader :span_id
+    # Span parent's span_id.
+    # @return [String]
+    attr_reader :parent_span_id
+
+    def initialize(scope)
+      @scope = scope
+      @trace_id = SecureRandom.uuid.delete("-")
+      @span_id = SecureRandom.uuid.delete("-").slice(0, 16)
+      @parent_span_id = nil
+      @baggage = nil
+    end
+
+    # Returns the trace context that can be used to embed in an Event.
+    # @return [Hash]
+    def get_trace_context
+      {
+        trace_id: trace_id,
+        span_id: span_id,
+        parent_span_id: parent_span_id
+      }
+    end
+
+    # Returns the sentry-trace header from the propagation context.
+    # @return [String]
+    def get_traceparent
+      "#{trace_id}-#{span_id}"
+    end
+
+    # Returns the W3C baggage header from the propagation context.
+    # @return [String, nil]
+    def get_baggage
+      populate_head_baggage if @baggage.nil? || @baggage.mutable
+      @baggage
+    end
+
+    # Returns the Dynamic Sampling Context from the baggage.
+    # @return [String, nil]
+    def get_dynamic_sampling_context
+      get_baggage&.dynamic_sampling_context
+    end
+
+    private
+
+    def populate_head_baggage
+      return unless Sentry.initialized?
```

Comment:
Since `PropagationContext` could only be initialised by an active `Scope`, which means the SDK is initialised, I think this is not necessary?

Reply:
this is again that multi-threaded case where one thread calls `Sentry.close` and another thread is still running this code. If we don't want this check, we will have to pass `configuration` into the constructor and not access it with `Sentry.configuration`. 

I'm ok with having these checks and using the global methods, this is also how python works.

Reply:
Since we have specific case in mind for the check, could we have a comment for it and perhaps also a test? Also, doesn't that mean anything that calls `Sentry.configuration` would need such check too, like [here](https://github.com/getsentry/sentry-ruby/blob/8abb2d213fb96ed45130956a3452e9cc7a097de0/sentry-ruby/lib/sentry/breadcrumb/sentry_logger.rb#L80)?

And perhaps we should implement a better version of `close` and `Sentry.initialize?` 🤔 

Specifically:

- We don't erase hub variables, `@background_worker` and `@session_flusher`
- `Sentry.initialize?` is backed by a simple boolean state that's changed by `Sentry.ini` and `Sentry.close`.

We would still need `Sentry.initialize?` in some places to stop some actions after the SDK is closed, but we don't need it for preventing`nil` value anymore, which is getting out of hand IMO.

Reply:
yea I agree we should fix the close/initialize behavior but that's out of scope here, I'll make a separate issue.
But for the record, other SDKs also work like this, python has `hub/client` checks at the beginning of most functions.

Reply:
https://github.com/getsentry/sentry-ruby/issues/2096